### PR TITLE
b2b discount refactoring

### PR DESCRIPTION
**Pull request by vadym-uskov**
- refactored b2b discount tests (tests work only on stage now: https://glue.de.b2b.internal-testing.demo-spryker.com/   https://backoffice.de.b2b.internal-testing.demo-spryker.com/)
- added new method 'Response body parameter should be in range:'
- updated documentation

### CODE REVIEW

File: tests/api/b2b/glue/cart_endpoints/items/positive.robot [hunk 1]
```diff
@@ -279,6 +276,68 @@ Add_product_with_options_to_cart
...
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response body parameter should be:    [data][id]    ${cart_uid}
+    And Response body parameter should be:    [data][type]    carts
+    And Response body parameter should be:    [data][attributes][priceMode]    ${gross_mode}
+    And Response body parameter should be:    [data][attributes][currency]    ${currency_code_eur}
+    And Response body parameter should be:    [data][attributes][store]    ${store_de}
+    And Response body parameter should be:    [data][attributes][name]    Cart-${random}
+    And Response body parameter should be:    [data][attributes][isDefault]    True
+    And Response body parameter should be:    [data][attributes][totals][expenseTotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][discountTotal]    0
```
Comment (OKrivtsova) [inline tests/api/b2b/glue/cart_endpoints/items/positive.robot:296]: don't we know the total amount of discount?

File: tests/api/b2b/glue/checkout_endpoints/checkout/positive.robot [hunk 2]
```diff
@@ -331,4 +331,78 @@ Create_order_with_same_items_in_different_shipments
     And Response body parameter should be greater than:    [included][0][attributes][shipments][0][defaultGrossPrice]    0
     And Response body parameter should be:    [included][0][attributes][shipments][0][defaultNetPrice]    0
     And Response body parameter should be:    [included][0][attributes][shipments][0][currencyIsoCode]    ${currency_code_eur}
-    And Response should contain certain number of values:    [included][0][attributes][expenses]    idShipment    1
\ No newline at end of file
+    And Response should contain certain number of values:    [included][0][attributes][expenses]    idShipment    1
+
+Create_order_with_2_product_discounts
```
Comment (OKrivtsova) [inline tests/api/b2b/glue/checkout_endpoints/checkout/positive.robot:336]: please check if vouchers and exclusive discounts are integrated in b2b and if they are present in demo data. if yes - please add tests for them as well

File: resources/common/common_api.robot [hunk 3]
```diff
@@ -398,6 +398,33 @@ Response body parameter should be in:
     Log    ${data}
     Should Contain Any   ${data}    ${expected_value1}    ${expected_value2}    ${expected_value3}    ${expected_value4}    ignore_case=True
 
+Response body parameter with rounding should be:
+    [Documentation]    This keyword checks that the response saved  in ``${response_body}`` test variable contains the speficied parameter ``${json_path}`` that has rounding from -1 to +1.
```
Comment (OKrivtsova) [inline resources/common/common_api.robot:402]: better phrase: that was rounded and can differ from the expected value by 1 more cent or 1 less cent

File: resources/common/common_api.robot [hunk 4]
```diff
@@ -398,6 +398,31 @@ Response body parameter should be in:
...
+    ...    ``Response body parameter should be in range:    [data][attributes][discounts][0][amount]    ${discount_concrete_product_3_discount_2_unit_amount}    -1    1``
+    [Arguments]    ${json_path}    ${expected_value}    ${range_value_min}    ${range_value_max}
+    ${data}=    Get Value From Json    ${response_body}    ${json_path}
+    ${data}=    Convert To String    ${data}
+    ${data}=    Replace String    ${data}    '   ${EMPTY}
+    ${data}=    Replace String    ${data}    [   ${EMPTY}
+    ${data}=    Replace String    ${data}    ]   ${EMPTY}
+    Log    ${data}
+    ${range_value_min}=    Evaluate    ${expected_value} + ${range_value_min}
+    ${range_value_max}=    Evaluate    ${expected_value} + ${range_value_max}
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/common/common_api.robot]: - fixed keyword with hardcode values for range: -1 and +1
- keyword name was changed to 'Response body parameter with rounding should be:'

File: resources/environments/environments_api_b2b.json [hunk 5]
```diff
@@ -261,7 +253,22 @@
...
+        "discount_concrete_product_1_discount_2_unit_amount": "4240",
+        "discount_concrete_product_2_sku": "213216",
+        "discount_concrete_product_2_name": "office akktiv base cabinet - HxWxD 942 x 913 x 420 mm - light gray, RAL 7035",
+        "individual_concrete_product_2_discount_amount": "8096",
+        "discount_concrete_product_2_discount_1_unit_amount": "5783",
+        "discount_concrete_product_2_discount_2_unit_amount": "2313",
+        "discount_concrete_product_3_sku": "401768",
+        "discount_concrete_product_3_name": "EUROKRAFT aluminum hand truck - Load capacity 200 kg, with skids - load capacity 200 kg",
+        "discount_concrete_product_3_discount_1_unit_amount": "16383",
+        "discount_concrete_product_3_discount_2_unit_amount": "1882"
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/environments/environments_api_b2b.json]: Renamed variables to the format like: 'discount_amount_for_product_1_with_discount_20%_off_storage'

File: resources/common/common_api.robot [hunk 6]
```diff
@@ -398,6 +398,31 @@ Response body parameter should be in:
     Log    ${data}
     Should Contain Any   ${data}    ${expected_value1}    ${expected_value2}    ${expected_value3}    ${expected_value4}    ignore_case=True
 
+Response body parameter should be in range:
+    [Documentation]    This keyword checks that the response saved  in ``${response_body}`` test variable contains the speficied parameter ``${json_path}`` with the value that matches of the parameter ``${expected_value1}`` in custom range with  ``${range_value_min}``, ``${range_value_max}``.
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/common/common_api.robot]: fixed

File: resources/environments/environments_api_b2b.json [hunk 7]
```diff
@@ -117,14 +117,6 @@
...
     },
 
-    "shopping-list":{
-        "shopping_list_name": "My Shopping List",
-        "yves_user_1st_shopping_list_id": "cf032865-d1ad-5e27-803a-423bd15ced66",
-        "yves_user_1st_shopping_list_name": "Newcomers",
-        "yves_user_2nd_shopping_list_name": "Weekly office",
-        "yves_user_3rd_shopping_list_name": "Monthly office"
-    },
-
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/environments/environments_api_b2b.json:127]: it was duplicated and discussed with @kostiantynlitvinov

"shopping-list":{
        "shopping_list_name": "My Shopping List",
        "yves_user_1st_shopping_list_id": "cf032865-d1ad-5e27-803a-423bd15ced66",
        "yves_user_1st_shopping_list_name": "Newcomers",
        "yves_user_2nd_shopping_list_name": "Weekly office",
        "yves_user_3rd_shopping_list_name": "Monthly office"
    },

File: resources/environments/environments_api_b2b.json [hunk 8]
```diff
@@ -264,7 +256,22 @@
...
+        "discount_concrete_product_1_total_sum_of_discounts": "14840",
+        "discount_concrete_product_2_sku": "213216",
+        "discount_concrete_product_2_name": "office akktiv base cabinet - HxWxD 942 x 913 x 420 mm - light gray, RAL 7035",
+        "discount_concrete_product_2_price_with_discount_20%_off_storage": "5783",
+        "discount_concrete_product_2_price_with_discount_10%_off_minimum_order": "2313",
+        "discount_concrete_product_2_total_sum_of_discounts": "8096",
+        "discount_concrete_product_3_sku": "401768",
+        "discount_concrete_product_3_name": "EUROKRAFT aluminum hand truck - Load capacity 200 kg, with skids - load capacity 200 kg",
+        "discount_concrete_product_3_price_with_discount_20%_off_storage": "16383",
+        "discount_concrete_product_3_price_with_discount_10%_off_minimum_order": "1882"
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/environments/environments_api_b2b.json]: renamed variables to the format like: 'discount_amount_for_product_1_with_discount_20%_off_storage'

File: resources/environments/environments_api_b2b.json [hunk 9]
```diff
@@ -349,6 +356,13 @@
         "review_text": "This is review text",
         "default_review_rating": "5",
 
+        "discount_1_name": "20% off storage",
+        "discount_1_description": "Storage sale: 20% off all office storage products",
+        "discount_1_total_sum_for_discount_concrete_product_1_and_2": "16383",
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/environments/environments_api_b2b.json:361]: fixed

File: resources/environments/environments_api_b2b.json [hunk 10]
```diff
@@ -349,6 +356,13 @@
...
+        "discount_2_name": "10% off minimum order",
+        "discount_2_description": "Get a 10% discount on all orders above a certain value depending on currency and gross/net price mode",
+        "discount_2_total_sum_for_discount_concrete_product_1_and_2_and_3": "8435",
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply resources/environments/environments_api_b2b.json:364]: fixed

File: tests/api/b2b/glue/cart_endpoints/carts/positive.robot [hunk 11]
```diff
@@ -104,6 +131,78 @@ Get_cart_by_cart_id_with_included_items
...
+    And Response body parameter should be:    [data][attributes][store]    ${store_de}
+    And Response body parameter should be:    [data][attributes][name]    ${test_cart_name}-${random}
+    And Response body parameter should be:    [data][attributes][isDefault]    True
+    #totals
+    And Response body parameter should be:    [data][attributes][totals][expenseTotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][discountTotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][taxTotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][subtotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][grandTotal]    0
+    And Response body parameter should be greater than:    [data][attributes][totals][priceToPay]    0
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply tests/api/b2b/glue/cart_endpoints/carts/positive.robot:161]: - added new method 'Response body parameter should be calculated:'

File: tests/api/b2b/glue/cart_endpoints/items/positive.robot (hunk 1 above)
Comment (vadym-uskov) [reply tests/api/b2b/glue/cart_endpoints/items/positive.robot:296]: fixed

File: tests/api/b2b/glue/cart_endpoints/items/positive.robot [hunk 12]
```diff
@@ -279,6 +276,68 @@ Add_product_with_options_to_cart
...
+    Then Response status code should be:    201
+    And Response reason should be:    Created
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response body parameter should be:    [data][id]    ${cart_uid}
+    And Response body parameter should be:    [data][type]    carts
+    And Response body parameter should be:    [data][attributes][priceMode]    ${gross_mode}
+    And Response body parameter should be:    [data][attributes][currency]    ${currency_code_eur}
+    And Response body parameter should be:    [data][attributes][store]    ${store_de}
+    And Response body parameter should be:    [data][attributes][name]    Cart-${random}
+    And Response body parameter should be:    [data][attributes][isDefault]    True
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply tests/api/b2b/glue/cart_endpoints/items/positive.robot:294]: fixed

File: resources/common/common_api.robot (hunk 3 above)
Comment (vadym-uskov) [reply resources/common/common_api.robot:402]: fixed

File: tests/api/b2b/glue/order_endpoints/orders/positive.robot [hunk 13]
```diff
@@ -444,6 +444,83 @@ Get_order_by_order_id_with_split_shipment_&_include
     And Response body parameter should be:    [included][1][attributes][shippingAddress][iso2Code]    ${default_iso2Code}
     And Response body parameter should not be EMPTY:    [included][1][links]
 
+Get_order_by_order_id_with_2_product_discounts
```
[the remark this answers is not in the document]
Comment (vadym-uskov) [reply tests/api/b2b/glue/order_endpoints/orders/positive.robot:447]: - fixed
- added new method 'Perform arithmetical calculation with two arguments:'

File: tests/api/b2b/glue/checkout_endpoints/checkout/positive.robot (hunk 2 above)
Comment (vadym-uskov) [reply tests/api/b2b/glue/checkout_endpoints/checkout/positive.robot:336]: - added 'free_shipping_discount' tests

**Review by OKrivtsova — APPROVED** [review]

**Pull request merged**